### PR TITLE
chore(dao): Add explicit ON DELETE CASCADE for ownership

### PR DESCRIPTION
**Pull request by john-bodley**
### SUMMARY

Per  [SIP-92 Proposal for restructuring the Python code base](https://github.com/apache/superset/issues/20630) https://github.com/apache/superset/pull/24331 organized all the DAOs to be housed within a shared folder—the result of which highlighted numerous inconsistencies, repetition, and inefficiencies.

This PR is somewhat akin to https://github.com/apache/superset/pull/24488 where explicit `ON DELETE CASCADE` operations are set for the ownership associations.

### TESTING INSTRUCTIONS

CI.

### ADDITIONAL INFORMATION

- [x] Has associated issue: https://github.com/apache/superset/issues/20630
- [ ] Required feature flags:
- [ ] Changes UI
- [x] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [x] Migration is atomic, supports rollback & is backwards-compatible
  - [x] Confirm DB migration upgrade and downgrade tested
  - [x] Runtime estimates and downtime expectations provided:  Takes ~ 1 minute for Airbnb's instance w/o load.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/daos/report.py [hunk 1]
```diff
@@ -95,30 +95,6 @@ def find_by_database_ids(database_ids: list[int]) -> list[ReportSchedule]:
...
-    @classmethod
-    def delete(
-        cls,
-        items: ReportSchedule | list[ReportSchedule],
-        commit: bool = True,
-    ) -> None:
-        item_ids = [item.id for item in get_iterable(items)]
-        try:
-            # Clean owners secondary table
-            report_schedules = (
```
Comment (john-bodley) [inline superset/daos/report.py:107]: I'm not really sure why the reports were refetched as opposed to using the `items` provides. Note after removing the need to remove the owners this logic simply reduced to the `BaseDAO.delete` method and thus is unnecessary.

File: superset/examples/birth_names.py [hunk 2]
```diff
@@ -535,7 +535,6 @@ def create_dashboard(slices: list[Slice]) -> Dashboard:
     dash = db.session.query(Dashboard).filter_by(slug="births").first()
     if not dash:
         dash = Dashboard()
-        dash.owners = []
```
Comment (john-bodley) [inline superset/examples/birth_names.py:538]: No need to initialize this.

File: superset/migrations/versions/2023-06-22_13-39_6fbe660cac39_add_on_delete_cascade_for_tables_references.py [hunk 3]
```diff
@@ -21,74 +21,45 @@
 Create Date: 2023-06-22 13:39:47.989373
 
 """
-from __future__ import annotations
 
 # revision identifiers, used by Alembic.
 revision = "6fbe660cac39"
 down_revision = "90139bf715e4"
 
-import sqlalchemy as sa
```
Comment (john-bodley) [inline superset/migrations/versions/2023-06-22_13-39_6fbe660cac39_add_on_delete_cascade_for_tables_references.py:30]: Same logic as before, but using the more generalized shared constraints logic.

File: superset/tables/models.py [hunk 4]
```diff
@@ -53,12 +53,12 @@
     Model.metadata,  # pylint: disable=no-member
     sa.Column(
         "table_id",
-        sa.ForeignKey("sl_tables.id", ondelete="cascade"),
+        sa.ForeignKey("sl_tables.id", ondelete="CASCADE"),
```
Comment (john-bodley) [inline superset/tables/models.py:56]: Using uppercase for consistency.

File: tests/integration_tests/charts/commands_tests.py [hunk 5]
```diff
@@ -282,8 +282,6 @@ def test_import_v1_chart(self, sm_g, utils_g):
 
         assert chart.owners == [admin]
 
-        chart.owners = []
-        database.owners = []
```
Comment (john-bodley) [inline tests/integration_tests/charts/commands_tests.py:286]: Databases don't have owners so this line was a red herring.

File: tests/integration_tests/commands_test.py [hunk 6]
```diff
@@ -146,9 +146,6 @@ def test_import_assets(self):
 
         assert dashboard.owners == [self.user]
 
-        dashboard.owners = []
-        chart.owners = []
-        database.owners = []
```
Comment (john-bodley) [inline tests/integration_tests/commands_test.py:151]: See above.

File: tests/integration_tests/commands_test.py [hunk 7]
```diff
@@ -190,10 +187,6 @@ def test_import_v1_dashboard_overwrite(self):
         chart = dashboard.slices[0]
         dataset = chart.table
         database = dataset.database
-        dashboard.owners = []
-
-        chart.owners = []
-        database.owners = []
```
Comment (john-bodley) [inline tests/integration_tests/commands_test.py:196]: See above.

File: tests/integration_tests/dashboards/commands_tests.py [hunk 8]
```diff
@@ -573,9 +573,6 @@ def test_import_v1_dashboard(self, sm_g, utils_g):
 
         assert dashboard.owners == [admin]
 
-        dashboard.owners = []
-        chart.owners = []
-        database.owners = []
```
Comment (john-bodley) [inline tests/integration_tests/dashboards/commands_tests.py:578]: See above.

File: tests/integration_tests/datasets/commands_tests.py [hunk 9]
```diff
@@ -400,7 +400,6 @@ def test_import_v1_dataset(self, sm_g, utils_g):
         assert column.description is None
         assert column.python_date_format is None
 
-        dataset.database.owners = []
```
Comment (john-bodley) [inline tests/integration_tests/datasets/commands_tests.py:403]: See above.

File: tests/integration_tests/datasets/commands_tests.py [hunk 10]
```diff
@@ -528,7 +527,6 @@ def test_import_v1_dataset_existing_database(self, mock_g):
         )
         assert len(database.tables) == 1
 
-        database.owners = []
```
Comment (john-bodley) [inline tests/integration_tests/datasets/commands_tests.py:531]: See above.

File: superset/migrations/versions/2023-07-07_15-51_6d05b0a70c89_add_on_delete_cascade_for_owners_references.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+Revision ID: 6d05b0a70c89
+Revises: 240d23c7f86f
+Create Date: 2023-07-07 15:51:57.964635
+
+"""
+
+# revision identifiers, used by Alembic.
+revision = "6d05b0a70c89"
+down_revision = "240d23c7f86f"
```
[the remark this answers is not in the document]
Comment (craig-rueda) [reply superset/migrations/versions/2023-07-07_15-51_6d05b0a70c89_add_on_delete_cascade_for_owners_references.py]: I would say that updating the filename is fine. Ideally the filename would be somewhat close to the date that the PR is merged as to keep things somewhat sorted.

**Review by craig-rueda — APPROVED** [review]
LGTM

File: superset/migrations/versions/2023-07-07_15-51_6d05b0a70c89_add_on_delete_cascade_for_owners_references.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (john-bodley) [reply superset/migrations/versions/2023-07-07_15-51_6d05b0a70c89_add_on_delete_cascade_for_owners_references.py]: Done.

**Comment by trinhtuanvy** [comment]
Thanks help fix

**Pull request merged**